src/Servers/HttpSys/src/RequestProcessing/RequestStream.cs

```diff
@@ -345,68 +246,74 @@ public override unsafe Task<int> ReadAsync(byte[] buffer, int offset, int size,
                 cancellationRegistration = RequestContext.RegisterForCancellation(cancellationToken);
             }
 
-            asyncResult = new RequestStreamAsyncResult(this, null, null, buffer, offset, dataRead, cancellationRegistration);
-            uint bytesReturned;
+            uint dataRead = 0;
+            uint statusCode = 0;
+
+            // Pin the user buffer
+            var handle = buffer.Pin();
 
             try
             {
-                uint flags = 0;
-
-                statusCode =
-                    HttpApi.HttpReceiveRequestEntityBody(
-                        RequestQueueHandle,
-                        RequestId,
-                        flags,
-                        asyncResult.PinnedBuffer,
-                        (uint)size,
-                        out bytesReturned,
-                        asyncResult.NativeOverlapped);
+                using (cancellationRegistration)
+                {
+                    (statusCode, dataRead) = await ReadEntityBodyAsync(handle, buffer.Length);
```

Comment:
Why did this become a separate method, to limit the size of ReadAsyncCore?

Doesn't the extra async/await state machine here add overhead?

Reply:
Because we need to run logic after the await. We can use the trick Kestrel uses and avoid the state machine in the case where it completes synchronously.